### PR TITLE
automated test during payment

### PR DESCRIPTION
**Pull request by fmessina-reef**
# Changes
Previously a paddle test would have required a manual interaction, this has been automated.

## new configuration

When running the paddle test you will need to set the `PADDLE_TEST_EMAIL` env variable, the test should alert you when this has not been done.

### CODE REVIEW

File: demo/demo/settings.py [hunk 1]
```diff
@@ -105,6 +105,9 @@
     }
 }
 
+# paddle test email.
```
Comment (agoncharov-reef) [inline demo/demo/settings.py]: This comment is redundant :)

File: demo/demo/settings.py [hunk 2]
```diff
@@ -105,6 +105,9 @@
...
+PADDLE_TEST_EMAIL = environ.get('PADDLE_TEST_EMAIL', '')
```
Comment (agoncharov-reef) [inline demo/demo/settings.py]: This should be placed near other `PADDLE_...` variables, not separately. Also, having defaults for env vars in `settings.py` is [forbidden](https://github.com/reef-technologies/handbook/#74-restricted-patterns--forbidden-technologies)

File: demo/demo/tests/test_paddle.py [hunk 3]
```diff
@@ -36,8 +174,7 @@ def test__payment_flow__regular(paddle, user_client, plan, card_number):
 
     assert 'payment_url' in payment.metadata
 
-    # TODO: automate this
-    input(f'Use card {card_number} to pay here: {redirect_url}\nThen press Enter')
+    automate_payment(redirect_url, card_number)
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Forgot to mention but there is other place where this should be applied - it also uses `input()`: https://github.com/reef-technologies/django-subscriptions-rt/blob/master/demo/demo/tests/test_paddle.py#L163

File: demo/demo/tests/test_paddle.py [hunk 4]
```diff
@@ -17,6 +20,141 @@
 from subscriptions.utils import fromisoformat
 
 
+def automate_payment(url, card):
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: For some weird reason I get three receipts instead of one, looks like you're submitting some form multiple times.
Screenshot_2023-07-18_15-37-36

File: demo/demo/tests/test_paddle.py [hunk 5]
```diff
@@ -17,6 +20,141 @@
...
+    """
+    This function replicates all the steps that are done at client level in the paddle payment page.
+    """
+    email = settings.PADDLE_TEST_EMAIL
+    assert email, 'Please configure the `PADDLE_TEST_EMAIL` environment variable.'
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: This functions takes card as an argument but at the same time takes email from settings. I would rather make the function "pure", i.e. not depending on anything except its arguments - so `email` should be an argument which would be passed from outer scope - the test.

File: demo/demo/tests/test_paddle.py [hunk 6]
```diff
@@ -17,6 +20,141 @@
...
+    # get the browser info token, which will be used later on.
+    browser_info = url.strip('/').split('/')[-1]
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: `maxsplit` avoids redundant splitting:
`url.removesuffix('/').rsplit('/', maxsplit=1)[-1]`

File: demo/demo/tests/test_paddle.py [hunk 7]
```diff
@@ -17,6 +20,141 @@
...
+    # hit the redirect url, and start the payment procedure
+    payment_page = session.get(url, allow_redirects=True)
+
+    # the checkout id has been initialized.
+    payment_page_query = parse.parse_qs(parse.urlsplit(payment_page.url).query)
+    checkout_id = payment_page_query['checkout_id'][0]
+    assert 'checkout_id' in payment_page_query
+
+    # parse the react to get the needed urls
+    checkout_api_url = payment_page.text.split("REACT_APP_CHECKOUT_API_URL: '")[-1].split("',")[0]
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: So you're splitting a big html into two chunks by `REACT_APP_...` string, then splitting the second chunk into two chunks by `',`. This is quite wasteful, why not using simple regex which is more readable anyway?
```
ipdb> re.search(r"REACT_APP_CHECKOUT_API_URL: '(.+)',", payment_page.text)[1]
'https://sandbox-checkout-service.paddle.com'

```

File: demo/demo/tests/test_paddle.py [hunk 8]
```diff
@@ -17,6 +20,141 @@
...
+    browser_info = url.strip('/').split('/')[-1]
+    session = requests.Session()
+
+    # hit the redirect url, and start the payment procedure
+    payment_page = session.get(url, allow_redirects=True)
+
+    # the checkout id has been initialized.
+    payment_page_query = parse.parse_qs(parse.urlsplit(payment_page.url).query)
+    checkout_id = payment_page_query['checkout_id'][0]
+    assert 'checkout_id' in payment_page_query
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Wait, how is it possible to fail this assertion?

File: demo/demo/tests/test_paddle.py [hunk 9]
```diff
@@ -17,6 +20,141 @@
...
+
+    # parse the react to get the needed urls
+    checkout_api_url = payment_page.text.split("REACT_APP_CHECKOUT_API_URL: '")[-1].split("',")[0]
+    ld_proxy_domain = payment_page.text.split("REACT_APP_LD_PROXY_URL: '")[-1].split("',")[0]
+    spreedly_api_url = payment_page.text.split("REACT_APP_SPREEDLY_API_URL: '")[-1].split("',")[0]
+
+    # Sanity checks on the variables
+    assert 'paddle' in checkout_api_url
+    checkout_api_url = f'{checkout_api_url}/checkout/{checkout_id}'
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: `checkout_api_url += f'/checkout/{checkout_id}'`

File: demo/demo/tests/test_paddle.py [hunk 10]
```diff
@@ -17,6 +20,141 @@
...
+    if not checkout_api_url.endswith('/'):
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: I think since it ends with `checkout_id`, it will never end with `/`

File: demo/demo/tests/test_paddle.py [hunk 11]
```diff
@@ -17,6 +20,141 @@
...
+        checkout_api_url += '/'
+
+    assert 'paddle' in ld_proxy_domain
+
+    assert 'spreedly' in spreedly_api_url
+    assert spreedly_api_url.endswith('v1'), 'Unexpected version of the spreedly API, this test might not work correctly'
+    if not spreedly_api_url.endswith('/'):
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: these are mutually exclusive

File: demo/demo/tests/test_paddle.py [hunk 12]
```diff
@@ -17,6 +20,141 @@
...
+        spreedly_api_url += '/'
+
+    # prepare the customer info request
+    customer_info_url = parse.urljoin(checkout_api_url, 'customer-info')
+    customer_info = requests.post(
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py:58]: Response status code not checked

File: demo/demo/tests/test_paddle.py [hunk 13]
```diff
@@ -17,6 +20,141 @@
...
+    )
+
+    # fetch the CARD payment method
+    for payment_method in payment_methods:
+        if payment_method['type'] == 'CARD':
+            assert 'spreedly_options' in payment_method
+            assert 'spreedly_environment_key' in payment_method['spreedly_options']
+            break
+    else:
+        assert False, 'Unexpected response from Paddle, payment method not found'
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]:
```python
from more_itertools import one

payment_method = one(method for method in payment_methods if method['type'] == 'CARD')
```

File: demo/demo/tests/test_paddle.py [hunk 14]
```diff
@@ -17,6 +20,141 @@
...
+                'country_code': 'PL',
+                'postcode': 12345
+            }
+        }
+    )
+    payment_methods = customer_info.json().get(
+        'data', {}
+    ).get(
+        'available_payment_methods', []
+    )
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Since you really need this info later, I think you want it to fail as early as possible instead of providing defaults and catching exceptions at later stages:
```python
payment_methods = customer_info.json()['data']['available_payment_methods']
```

File: demo/demo/tests/test_paddle.py [hunk 15]
```diff
@@ -17,6 +20,141 @@
...
+
+    # fetch the CARD payment method
+    for payment_method in payment_methods:
+        if payment_method['type'] == 'CARD':
+            assert 'spreedly_options' in payment_method
+            assert 'spreedly_environment_key' in payment_method['spreedly_options']
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: No need to check it here, since if it doesn't exist then it will fail later when doing
```python
spreedly_environment_key = payment_method['spreedly_options']['spreedly_environment_key']
```

File: demo/demo/tests/test_paddle.py [hunk 16]
```diff
@@ -17,6 +20,141 @@
...
+    else:
+        assert False, 'Unexpected response from Paddle, payment method not found'
+
+    # Now we should have received the spreedly environment token.
+    spreedly_environment_key = payment_method['spreedly_options']['spreedly_environment_key']
+    assert spreedly_environment_key
+
+    # Set the payment method to CARD
+    set_payment_method_url = parse.urljoin(checkout_api_url, 'payment-method')
+    requests.patch(
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Response status not checked

File: demo/demo/tests/test_paddle.py [hunk 17]
```diff
@@ -17,6 +20,141 @@
...
+        set_payment_method_url,
+        json={
+            'data': {
+                "payment_method_type": "CARD"
+            }
+        }
+    )
+
+    # Send the card details to spreedly adn fetch the transaction token
+    spreedly_url = parse.urljoin(spreedly_api_url, f'payment_methods.json?environment_key={spreedly_environment_key}')
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Inconsistent. You're using `urljoin` to "join url parts the right way", however you're constructing query string with string interpolation instead of `urlencode`.

File: demo/demo/tests/test_paddle.py [hunk 18]
```diff
@@ -17,6 +20,141 @@
...
+    expire_date = date.today() + timedelta(90)
+    payment_response = session.post(spreedly_url, json={
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Response status not checked

File: demo/demo/tests/test_paddle.py [hunk 19]
```diff
@@ -17,6 +20,141 @@
...
+                "kind": "credit_card",
+                "month": expire_date.strftime('%m'),
+                "number": card,
+                "verification_value": "123",
+                "year": expire_date.strftime('%Y'),
+                "zip": "12345"
+            }
+        }
+    })
+    transaction_token = payment_response.json().get('transaction', {}).get('payment_method', {}).get('token')
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Again, you probably don't want to use `get` here, you don't need defaults

File: demo/demo/tests/test_paddle.py [hunk 20]
```diff
@@ -17,6 +20,141 @@
...
+    assert transaction_token, 'Unexpected response from spreedly'
+
+    # Send the payment information to paddle
+    make_payment_url = parse.urljoin(checkout_api_url, 'pay-card')
+    make_payment_response = session.post(make_payment_url, json={
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py:127]: Response status not checked

File: demo/demo/tests/test_paddle.py [hunk 21]
```diff
@@ -17,6 +20,141 @@
...
+            "three_d_s": {
+                "spreedly": {
+                    "browser_info": browser_info
+                }},
+            "token": transaction_token
+        }
+    })
+
+    # not sure if we need it, but I will still replicate the client behaviour
+    session.get(parse.urljoin(ld_proxy_domain, f'/sdk/evalx/{browser_info}'))
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Response status not checked

File: demo/demo/tests/test_paddle.py [hunk 22]
```diff
@@ -17,6 +20,141 @@
...
+
+    # 3D Secure - assuming it is needed
+    three_d_s_url = make_payment_response.json()['data']['three_d_s']['spreedly']['checkout_url']
+    three_d_s = session.get(three_d_s_url)
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py:146]: Response status not checked

File: demo/demo/tests/test_paddle.py [hunk 23]
```diff
@@ -17,6 +20,141 @@
...
+
+    # This page contains (your payment is successful, click here to return to the merchant)
+    # in general this step might be necessary to finalize the transaction.
+    finalize_three_d_s_url = three_d_s.text.split('href="')[-1].split('"')[0]
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Use regex

File: demo/demo/tests/test_paddle.py [hunk 24]
```diff
@@ -17,6 +20,141 @@
...
+    assert 'spreedly' in finalize_three_d_s_url
+    session.get(finalize_three_d_s_url)
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Response status not checked

File: demo/demo/tests/test_paddle.py [hunk 25]
```diff
@@ -17,6 +20,141 @@
...
+    session.post(parse.urljoin(checkout_api_url, 'pay-card/3ds-finished'))
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Response status not checked

File: demo/demo/tests/test_paddle.py [hunk 26]
```diff
@@ -17,6 +20,141 @@
...
+    This function replicates all the steps that are done at client level in the paddle payment page.
+    """
+    email = settings.PADDLE_TEST_EMAIL
+    assert email, 'Please configure the `PADDLE_TEST_EMAIL` environment variable.'
+    # get the browser info token, which will be used later on.
+    browser_info = url.strip('/').split('/')[-1]
+    session = requests.Session()
+
+    # hit the redirect url, and start the payment procedure
+    payment_page = session.get(url, allow_redirects=True)
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py:34]: Response status not checked. 
```python
payment_page.raise_for_status()
```

**Review by agoncharov-reef — CHANGES_REQUESTED** [review]
Overall it does the job, but receiving 3 receipts with different transaction IDs makes me think something goes wrong. As far as I remember, manual payment produces 1 receipt - please verify.

File: demo/demo/settings.py (hunk 1 above)
Comment (fmessina-reef) [reply demo/demo/settings.py]: ok

File: demo/demo/settings.py (hunk 2 above)
Comment (fmessina-reef) [reply demo/demo/settings.py]: OK I did not think that it would have included not meaningfull defaults, I will need to document the variable to the readme among the exports.

File: demo/demo/tests/test_paddle.py (hunk 4 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: Ok this is a bit weird I might have not notice because I ran it multiple times, let me fix it

File: demo/demo/tests/test_paddle.py (hunk 5 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: yeah it makes sense

File: demo/demo/tests/test_paddle.py (hunk 4 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: I inspect the code and it seems that they are related to `payment.subscription.charge_offline()` which triggers a new payment and a new email, it seems like we have the same behavior in the "manual" case.

File: demo/demo/tests/test_paddle.py (hunk 8 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: It should have been the other way aroind indeed

File: demo/demo/tests/test_paddle.py (hunk 7 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: ok

File: demo/demo/tests/test_paddle.py (hunk 11 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: right

File: demo/demo/tests/test_paddle.py (hunk 10 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 9 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 12 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py:58]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 13 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 14 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 15 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 16 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 17 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 18 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 19 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 21 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 22 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py:146]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 23 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 24 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 25 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: Removed s it does not seem needed

File: demo/demo/tests/test_paddle.py (hunk 26 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py:34]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 20 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py:127]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 3 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: Ok, I had to change the test a bit and use `tenacity`

**Comment by fmessina-reef** [comment]
Yeah it seems that I rushed it a bit, I should have fixed all the changes requests 
@agoncharov-reef

File: demo/demo/tests/test_paddle.py (hunk 6 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: :+1:

File: demo/demo/tests/test_paddle.py (hunk 4 above)
Comment (agoncharov-reef) [reply demo/demo/tests/test_paddle.py]: Oh, sorry, you are 100% correct. My bad.

File: demo/demo/tests/test_paddle.py (hunk 13 above)
Comment (agoncharov-reef) [reply demo/demo/tests/test_paddle.py]: Actually I didn't know about `too_short` and `too_long` parameters, you made it even nicer than my version

File: demo/demo/tests/test_paddle.py (hunk 17 above)
Comment (agoncharov-reef) [reply demo/demo/tests/test_paddle.py]: Nice solution with `params`!

File: demo/demo/tests/test_paddle.py [hunk 27]
```diff
@@ -304,7 +304,8 @@ def test__payment_flow__trial_period(trial_period, paddle, user, user_client, pl
     subscription = user.subscriptions.first()
     assert subscription.start == subscription.end
     assert subscription.initial_charge_offset == trial_period
-    # input(f'Use card {card_number} to pay here: {redirect_url}\nThen press Enter')
+    assert subscription.initial_charge_offset == trial_period
+
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: Nope :)

File: demo/demo/tests/test_paddle.py [hunk 28]
```diff
@@ -247,13 +248,10 @@ def test__payment_flow__regular(paddle, user_client, plan, card_number, paddle_t
     payment.status = SubscriptionPayment.Status.PENDING
     payment.save()
 
-    # Retry a few times to give paddle the time to update the tranaction status
-    for attempt in Retrying(wait=wait_incrementing(start=2, increment=2), stop=stop_after_attempt(10)):
-        with attempt:
-            check_unfinished_payments(within=timedelta(hours=1))
-            payment = SubscriptionPayment.objects.latest()
-            if payment.status != SubscriptionPayment.Status.COMPLETED:
-                raise TryAgain()
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py]: But why is this deleted?

File: demo/demo/tests/test_paddle.py [hunk 29]
```diff
@@ -316,11 +313,14 @@ def test__payment_flow__trial_period(trial_period, paddle, user, user_client, pl
...
-    check_unfinished_payments(within=timedelta(hours=1))
-
-    payment = SubscriptionPayment.objects.latest()
+    # Retry a few times to give paddle the time to update the tranaction status
+    for attempt in Retrying(wait=wait_incrementing(start=2, increment=2), stop=stop_after_attempt(20)):
+        with attempt:
+            check_unfinished_payments(within=timedelta(hours=1))
+            payment = SubscriptionPayment.objects.latest()
+            if payment.status != SubscriptionPayment.Status.COMPLETED:
+                raise TryAgain()
```
Comment (agoncharov-reef) [inline demo/demo/tests/test_paddle.py:317]: Since it's the same block of code used in `test__payment_flow__regular` and `test__payment_flow__trial_period`, you could extract it to separate function and put it into `helpers.py`

File: demo/demo/tests/test_paddle.py (hunk 28 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py]: Doh !

File: demo/demo/tests/test_paddle.py (hunk 29 above)
Comment (fmessina-reef) [reply demo/demo/tests/test_paddle.py:317]: Hi I removed one by mistake but these are two different ones (EDIT I confused the two for a duplicate myself)
```python
    # ---- test_payment_status_endpoint_post ----
    for attempt in Retrying(wait=wait_incrementing(start=2, increment=2), stop=stop_after_attempt(10)):
        with attempt:
            response = user_client.post(f'/api/payments/{payment.id}/')
            assert response.status_code == 200, response.content
            result = response.json()
            if result['status'] != 'completed':
                raise TryAgain()
```

and 
```python
    # Retry a few times to give paddle the time to update the tranaction status
    for attempt in Retrying(wait=wait_incrementing(start=2, increment=2), stop=stop_after_attempt(20)):
        with attempt:
            check_unfinished_payments(within=timedelta(hours=1))
            payment = SubscriptionPayment.objects.latest()
            if payment.status != SubscriptionPayment.Status.COMPLETED:
                raise TryAgain()
```

I'm not sure whether it would make sense to put the automate_payment function in the `helpers.py` module as it exposes logic that seems exclusively used by paddle.

Comment (agoncharov-reef) [reply demo/demo/tests/test_paddle.py:317]: I agree, you are right

**Review by agoncharov-reef — APPROVED** [review]
